volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -105,40 +105,62 @@ def get_sections(self):
             yield attr
 
     def get_symbols(self):
-        if symbols.symbol_table_is_64bit(self._context, self.get_symbol_table().name):
+        """Get module symbols"""
+        if symbols.symbol_table_is_64bit(self._context, self.get_symbol_table_name()):
             prefix = "Elf64_"
         else:
             prefix = "Elf32_"
         elf_table_name = intermed.IntermediateSymbolTable.create(
-            self.context,
-            self.config_path,
+            self._context,
+            self._context.modules["kernel"].config_path,
             "linux",
             "elf",
             native_types=None,
             class_types=elf.class_types,
         )
 
         syms = self._context.object(
-            self.get_symbol_table().name + constants.BANG + "array",
+            self.get_symbol_table_name() + constants.BANG + "array",
             layer_name=self.vol.layer_name,
             offset=self.section_symtab,
             subtype=self._context.symbol_space.get_type(
                 elf_table_name + constants.BANG + prefix + "Sym"
             ),
             count=self.num_symtab + 1,
         )
+
         if self.section_strtab:
             for sym in syms:
-                sym.set_cached_strtab(self.section_strtab)
-                yield sym
+                try:
+                    sym_offset = self.section_strtab + sym.st_name
+                    sym_name = self._context.layers[self.vol.layer_name].read(
+                        sym_offset, 512
+                    )  # 512 is the value of KSYM_NAME_LEN
+                except exceptions.PagedInvalidAddressException:
+                    continue
+
+                # Stop at first null byte (strtab is a null terminated strings list)
+                sym_name = sym_name.split(b"\x00")[0].decode("latin-1")
+                if sym_name != "":
+                    # Normalize sym_value
+                    mask = self._context.layers[self.vol.layer_name].address_mask
+                    sym_value = sym.st_value & mask
+                    yield (sym_name, sym_value, sym_offset)
```

Comment:
This is now yielding a totally different type of result, which is a hard API break and would need version number bumping.  It isn't clear wht this information is being returned separately rather than as a symbol object as it had been.

Reply:
Regarding the fact that the previous implementation, using an undocumented set_cached_strtab function, wasn't working with pretty obvious errors, and with no GitHub issues raised, I started noticing that this code didn't seem to be used anywhere in the codebase.

Doing so, to follow the style of previous Volatility2 implementation of this [section](https://github.com/volatilityfoundation/volatility/blob/master/volatility/plugins/overlays/linux/linux.py#L994), I decided to correct the logic needed to read symbols :

iterate over the "ElfXX_Sym" structures stored in "strtab"
extract each "st_name", which is an offset inside the "strtext" string list
read [512 bytes](https://elixir.bootlin.com/linux/v6.6/source/include/linux/kallsyms.h#L18) starting from "strtext" offset + "st_name"
check if read was successful, and if symbol isn't empty
The return value isn't a "sym" object, because it does not provide any useful information for the caller, instead I adapted :

# https://github.com/volatilityfoundation/volatility/blob/master/volatility/plugins/overlays/linux/linux.py#L1015
if sym_name != "":
                ret_syms.append((str(sym_name), sym_struct.st_value.v()))
I also added the absolute symbol offset to the tuple returned, but one can say that is not really interesting, and we could stick with (symbol_name, symbol_value).

Finally, I think this would eventually result in an API break, because right now I do not see how it could be usable (and where it is used as is) ?

Reply:
Sadly, that doesn't matter too much.  If we exposed it, then it's part of the API.  So the options then become:

* Add something extra under a different name that does the right thing (bump the minor number)
* Change this function so it returns the right thing in the same format as the previous one did (bump the patch number)
* Remove the old one entirely and make a new one (or just change this one) with a different API, whilst bumping the major version number of whatever component this is associated with. (bump the major number)

We shouldn't be afraid of bumping version numbers, it just needs to make us think carefully about our API when we design it to avoid doing it too often and confusing everybody.  I have no problem with the options or with bumping the numbers appropriately, and if it isn't used much it should be fine, but we're not the only users of the library, so if we are going to make a change we need to stick to our own rules about making changes...  5:)